pkg/virt-handler/rest/console.go

```diff
@@ -45,17 +46,75 @@ type ConsoleHandler struct {
 	serialLock           *sync.Mutex
 	vncLock              *sync.Mutex
 	vmiInformer          cache.SharedIndexInformer
+	usbredir             [v1.UsbredirMaxNumberOfDevices]UsbredirHandler
+}
+
+type UsbredirHandler struct {
+	lock      *sync.Mutex
+	stopChans map[types.UID](chan struct{})
+	inUse     bool
 }
 
 func NewConsoleHandler(podIsolationDetector isolation.PodIsolationDetector, vmiInformer cache.SharedIndexInformer) *ConsoleHandler {
-	return &ConsoleHandler{
+	c := &ConsoleHandler{
 		podIsolationDetector: podIsolationDetector,
 		serialStopChans:      make(map[types.UID](chan struct{})),
 		vncStopChans:         make(map[types.UID](chan struct{})),
 		serialLock:           &sync.Mutex{},
 		vncLock:              &sync.Mutex{},
 		vmiInformer:          vmiInformer,
 	}
+
+	for i := 0; i < v1.UsbredirMaxNumberOfDevices; i++ {
+		c.usbredir[i] = UsbredirHandler{
+			lock:      &sync.Mutex{},
+			stopChans: make(map[types.UID](chan struct{})),
+			inUse:     false,
+		}
+	}
+
+	return c
+}
+
+func (t *ConsoleHandler) USBRedirHandler(request *restful.Request, response *restful.Response) {
+	vmi, code, err := getVMI(request, t.vmiInformer)
+	if err != nil {
+		log.Log.Object(vmi).Reason(err).Error("Failed to retrieve VMI")
+		response.WriteError(code, err)
+		return
+	}
+
+	// Get the next free usbredir socket
+	slot := -1
+	for i, handler := range t.usbredir {
+		if !handler.inUse {
+			slot = i
+			break
+		}
+	}
+	if slot < 0 {
+		log.Log.Object(vmi).Reason(err).Error("All USB Redir devices are being used")
+		response.WriteError(http.StatusBadRequest, err)
+		return
+	}
+	handler := &t.usbredir[slot]
+
+	// Flag as used even if socket does not exist as a workaround to not handling how many
+	// sockets were created.
+	handler.inUse = true
```

Comment:
The use of `inUse` is not very clear to me. You have a lock on the slot, why do we need the flag as well?
Also, shouldn't you release it if this function returns? If we enter the error below, wouldn't the slot remain locked?

Reply:
To idea is to handle when you have multiple usbredir channels that can be used. We need to track what are in use and what are free to be used. The flag here does that. The comment points out a workaround that I'm setting it to true before checking it the socket exists - that mean we would not check it again in the future as it would stay true.

Let me know if you see a way to avoid this workaround, that is, I would like to know how many /run/kubevirt-private/.../virt-usbredir$d were created.

I'm not sure how can I replace the inUse flag with the lock. The lock seems to be used only to avoid multiple requests to the same resource... I think this code can be improved a bit still, but I'm actually new to go.

Thanks for the review!